src/jdk.internal.vm.ci/share/classes/jdk/vm/ci/hotspot/HotSpotJVMCIRuntime.java

```diff
@@ -675,11 +690,27 @@ synchronized HotSpotResolvedObjectTypeImpl fromMetaspace(long klassPointer) {
         if (javaType == null) {
             String name = compilerToVm.getSignatureName(klassPointer);
             javaType = new HotSpotResolvedObjectTypeImpl(klassPointer, name);
-            resolvedJavaTypes.put(klassPointer, new WeakReference<>(javaType));
+            resolvedJavaTypes.put(klassPointer, new KlassWeakReference(klassPointer, javaType, resolvedJavaTypesQueue));
         }
+        expungeStaleEntries();
         return javaType;
     }
 
+
+    /**
+     * Clean up WeakReferences whose referents have been cleared.
+     */
+    private void expungeStaleEntries() {
+        KlassWeakReference current = (KlassWeakReference) resolvedJavaTypesQueue.poll();
+        while (current != null) {
+            // Make sure the entry is still mapped to the weak reference
```

Comment:
The atomicity of this test-and-update relies on `fromMetaspace` being synchronized right? Maybe worth pointing this out in the comment.